File: web/js/featurize.mjs
// Port of app/backend/featurize.py Featurizer. Builds the exact 64-d prediction-time feature vector
// per perturbation from the bundled reference tables. Writing into a Float32Array gives the float32
// semantics of Python's np.array(rows, dtype=float32) for free.

export const CONDS = ["Rest", "Stim8hr", "Stim48hr"];
export const EMB_DIM = 50;

const NOTE =
  "Features are computed at inference from the bundled reference and match the training feature " +
  "construction. Contexts must be one of Rest, Stim8hr, Stim48hr; other contexts are out of " +
  "distribution for this model and are rejected. Genes with no reference entry are scored with " +
  "missing features (the tree handles them) but less reliably.";

// Python round() is round-half-to-even; matters only at exact 3rd-decimal halves but reproduce it.
function round3(x) {
  const scaled = x * 1000;
  const floor = Math.floor(scaled);
  const diff = scaled - floor;
  let r;
  if (diff > 0.5) r = floor + 1;
  else if (diff < 0.5) r = floor;
  else r = floor % 2 === 0 ? floor : floor + 1;
  return r / 1000;
}

export class Featurizer {
  // refs: { condIdx:Map, nGenes, stats:Float32Array, geneCol:Map<ens,col>, ensSet:Set,
  //         emb:Map<ens,Float32Array(50)>, sym2ens:Map, ens2sym:Map, sim:Map, neighbors:Map }
  constructor(refs) {
    this.r = refs;
  }

  _toEns(g) {
    g = String(g).trim();
    if (this.r.ensSet.has(g)) return g;
    return this.r.sym2ens.get(g) ?? null;
  }

  _stat(which, cond, col) {
    return this.r.stats[which * 3 * this.r.nGenes + cond * this.r.nGenes + col];
  }

  featurize(predictions) {
    const { condIdx, geneCol, emb, sim } = this.r;
    const rows = [], syms = [], unmapped = [], badCtx = [];
    for (const p of predictions) {
      const src = String(p.src), dst = String(p.dst);
      if (!condIdx.has(src) || !condIdx.has(dst)) { badCtx.push([p.perturbed_gene, src, dst]); continue; }
      const ens = this._toEns(p.perturbed_gene);
      if (ens == null) { unmapped.push(String(p.perturbed_gene)); continue; }
      const gi = geneCol.has(ens) ? geneCol.get(ens) : null;
      const si = condIdx.get(src), di = condIdx.get(dst);

      const f = new Float32Array(64);
      f[0] = p.pred_magnitude;
      let k = 1;
      for (const ci of [si, di]) for (const which of [0, 1, 2]) f[k++] = gi == null ? NaN : this._stat(which, ci, gi);
      for (const c of CONDS) { f[k++] = src === c ? 1 : 0; f[k++] = dst === c ? 1 : 0; }
      const e = emb.get(ens);
      for (let j = 0; j < EMB_DIM; j++) f[13 + j] = e ? e[j] : NaN;
      f[63] = sim.has(ens) ? sim.get(ens) : NaN;

      rows.push(f);
      syms.push(String(p.perturbed_gene));
    }
    const report = {
      n_input: predictions.length,
      n_featurized: rows.length,
      n_unmapped_genes: unmapped.length,
      unmapped_genes_sample: unmapped.slice(0, 20),
      n_bad_context: badCtx.length,
      supported_contexts: CONDS,
      mapped_fraction: round3(rows.length / Math.max(1, predictions.length)),
      note: NOTE,
    };
    return { X: rows, syms, report };
  }

  // Nearest reference genes in co-expression space (precomputed, self-skipped, symbols). Display only.
  neighbors(gene, k = 5) {
    const ens = this._toEns(gene);
    if (ens == null || !this.r.neighbors.has(ens)) return [];
    return this.r.neighbors.get(ens).slice(0, k);
  }
}

File: pertema/featurize.py
"""Featurize a user's perturbation predictions into the exact 64-d prediction-time feature vector the frozen
estimator expects, using the reference data bundled in the model artifact. The feature order matches the
training-time construction (build_transfer_features + training-set similarity) exactly, verified by
round-tripping the bundled example, so a user's features are on the same footing as the training features.

Leakage-safe: features use only prediction-time information (the user's predicted effect magnitude, the
control-state baseline features of the perturbed gene, its co-expression embedding, the context indicators,
and similarity to the reference gene set). No true post-perturbation effect enters.
"""
import os

import numpy as np
import pandas as pd
from sklearn.neighbors import NearestNeighbors

CONDS = ["Rest", "Stim8hr", "Stim48hr"]
EMB_DIM = 50


class Featurizer:
    def __init__(self, ref_dir):
        bz = np.load(os.path.join(ref_dir, "control_baseline.npz"))
        self.gene_col = {str(g): i for i, g in enumerate(bz["genes"])}
        self.cond_idx = {c: i for i, c in enumerate([str(c) for c in bz["conditions"]])}
        self.baseline, self.dropout, self.donor_var = bz["baseline"], bz["dropout"], bz["donor_var"]
        ez = np.load(os.path.join(ref_dir, "gene_embedding.npz"))
        self.emb = {str(g): v for g, v in zip(ez["gene_ids"], ez["embedding"])}
        gm = pd.read_csv(os.path.join(ref_dir, "gene_map.csv"), dtype=str)
        self.sym2ens = dict(zip(gm["symbol"], gm["ensembl"]))
        self.ens2sym = dict(zip(gm["ensembl"], gm["symbol"]))
        self.ens_set = set(self.gene_col) | set(self.emb)
        # training-set similarity: nearest OTHER embedded gene, precomputed for every embedded gene
        genes = np.array(list(self.emb)); X = np.array([self.emb[g] for g in genes])
        nn = NearestNeighbors(n_neighbors=2, algorithm="brute", n_jobs=8).fit(X)
        d, _ = nn.kneighbors(X)
        self._sim = {g: float(d[i, 1]) for i, g in enumerate(genes)}
        self._nn, self._nn_genes = nn, genes

    def _to_ens(self, g):
        g = str(g).strip()
        if g in self.ens_set:
            return g
        return self.sym2ens.get(g)          # symbol -> Ensembl, or None if unmappable

    def neighbors(self, gene, k=5):
        """F6 gene-network neighborhood: the k nearest reference genes to the perturbed gene in co-expression
        embedding space (symbols where mappable). Prediction-time only, from the bundled reference."""
        ens = self._to_ens(gene)
        if ens is None or ens not in self.emb:
            return []
        q = np.asarray(self.emb[ens], float)[None, :]
        d, ix = self._nn.kneighbors(q, n_neighbors=min(k + 1, len(self._nn_genes)))
        out = []
        for dist, j in zip(d[0], ix[0]):
            g = str(self._nn_genes[j])
            if g == ens:
                continue                     # skip self
            out.append(self.ens2sym.get(g, g))
            if len(out) >= k:
                break
        return out

    def _sim_of(self, ens):
        if ens in self._sim:
            return self._sim[ens]
        if ens in self.emb:
            return self._sim[ens]
        return np.nan                        # no embedding -> unknown similarity (xgboost handles NaN)

    def featurize(self, predictions):
        """predictions: list of dicts with perturbed_gene, src, dst, and pred_magnitude (mean abs predicted
        effect). Returns (X (n,64), symbols, report)."""
        rows, syms, unmapped, bad_ctx = [], [], [], []
        for p in predictions:
            ens = self._to_ens(p["perturbed_gene"])
            src, dst = str(p["src"]), str(p["dst"])
            if src not in self.cond_idx or dst not in self.cond_idx:
                bad_ctx.append((p["perturbed_gene"], src, dst)); continue
            if ens is None:
                unmapped.append(str(p["perturbed_gene"])); continue
            gi = self.gene_col.get(ens)
            f = [float(p["pred_magnitude"])]
            for tag in (src, dst):
                ci = self.cond_idx[tag]
                for arr in (self.baseline, self.dropout, self.donor_var):
                    f.append(float(arr[ci, gi]) if gi is not None else np.nan)
            for c in CONDS:                    # interleaved src/dst one-hot, matching training
                f.append(1.0 if src == c else 0.0)
                f.append(1.0 if dst == c else 0.0)
            e = self.emb.get(ens, np.full(EMB_DIM, np.nan))
            f.extend([float(x) for x in e])
            f.append(self._sim_of(ens))
            rows.append(f); syms.append(str(p["perturbed_gene"]))
        X = np.array(rows, dtype=np.float32) if rows else np.zeros((0, 64), np.float32)
        report = {
            "n_input": len(predictions), "n_featurized": len(rows),
            "n_unmapped_genes": len(unmapped), "unmapped_genes_sample": unmapped[:20],
            "n_bad_context": len(bad_ctx),
            "supported_contexts": CONDS,
            "mapped_fraction": round(len(rows) / max(1, len(predictions)), 3),
            "note": ("Features are computed at inference from the bundled reference and match the training "
                     "feature construction. Contexts must be one of Rest, Stim8hr, Stim48hr; other contexts "
                     "are out of distribution for this model and are rejected. Genes with no reference entry "
                     "are scored with missing features (the tree handles them) but less reliably."),
        }
        return X, syms, report


def default_featurizer():
    """Load the featurizer's reference data from the bundled model. Override with PERTEMA_MODEL_DIR."""
    here = os.path.dirname(os.path.abspath(__file__))
    base = os.environ.get("PERTEMA_MODEL_DIR") or os.path.join(
        here, "..", "app", "model", "pertema_model_v0.1.0")
    return Featurizer(os.path.join(base, "reference"))
